cloudinstall/charms/__init__.py

```diff
@@ -364,15 +356,33 @@ def watch_deploy(self):
                 self.ui.status_error_message(msg)
             time.sleep(10)
 
+    def filter_valid_relations(self):
+        deployed_charms_names = [x.charm_name for x in self.deployed_charms]
+        valid_relations = []
+        for c in self.deployed_charms:
+            for rel_a, rel_b in c.related:
+                rel_a_name = rel_a.split(":")[0]
+                rel_b_name = rel_b.split(":")[0]
+                if rel_a_name in deployed_charms_names and \
+                   rel_b_name in deployed_charms_names:
+                    valid_relations.append((rel_a, rel_b))
+        return valid_relations
+
     @utils.async
     def watch_relations_async(self):
         self.watch_relations()
 
     def watch_relations(self):
+        """
+        Handle charm relations
+        """
+        for charm in self.deployed_charms:
+            self.charm_relations_q.put(charm)
         log.debug("Starting charm relations watcher.")
         while not self.charm_relations_q.empty():
             try:
                 charm = self.charm_relations_q.get()
+                charm.related = self.filter_valid_relations()
```

Comment:
Also as a style point I don't really like the idea of altering the charm's related list. since it's set in the class definition, it seems like it's meant to be immutable

what if instead of this, we make `filter_valid_relations` take a list of relations to validate as an argument, and then give `filter_valid_relations` to the charm when we add it to the queue (maybe by setting `charm.relation_validation_function`, an ivar I just made up) , and then the charm will use that function to filter the relations it tries to make in `set_relations` ?